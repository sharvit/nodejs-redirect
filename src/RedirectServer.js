import express from 'express';
import { createRedirectMiddleware } from './createRedirectMiddleware';

/**
 * Express server that redirect all requests to a different URL
 */
export class RedirectServer {
  /**
   * Creates an instance of RedirectServer
   * @param {string}  to            URl to redirect requests.
   * @param {Boolean} [chain=false] Will chain the request path to the redirect url.
   */
  constructor(to, chain = false) {
    /**
     * URl to redirect requests.
     * @type {string}
     */
    this.to = to;
    /**
     * Will chain the request path to the redirect url.
     * @type {boolean}
     */
    this.chain = chain;

    this._setupExpressServer();
  }

  /**
   * Setup express server
   */
  _setupExpressServer() {
    const { to, chain } = this;

    /**
     * Express server
     * @type {Express}
     */
    this.expressServer = express();

    // create the redirection middleware
    const redirectionMiddleware = createRedirectMiddleware({ to, chain });

    // setup the redirection middleware on all requests
    this.expressServer.get('*', redirectionMiddleware);
  }

  /**
   * Start the server on a port, start listening
   * @param  {Number} port Port for running the server
   */
  listen(port) {
    this.expressServer.listen(port);
  }
}
