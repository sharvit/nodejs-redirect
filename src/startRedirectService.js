import { RedirectServer } from './RedirectServer';

/**
 * Start a redirection service
 * @param  {Object}  options               Options object.
 * @param  {String}  options.to            URl to redirect requests.
 * @param  {Boolean} [options.chain=false] Will chain the request path to the redirect url.
 * @param  {Number}  [options.port=3000]   Port to run the service.
 */
export const startRedirectService = ({ to, chain = false, port = 3000 }) => {
  // create the redirection server
  const server = new RedirectServer(to, chain);

  // run the server and start listen to the port
  server.listen(port);

  // Render some console log output
  console.log('Redirection server listening on port: ' + port);
  console.log('Redirecting all requests to: ' + to);
  console.log('Chaining requests url is: ' + (chain ? 'on' : 'off'));
};
