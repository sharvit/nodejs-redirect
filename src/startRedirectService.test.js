import { RedirectServer } from './RedirectServer';
import { startRedirectService } from './startRedirectService';

jest.mock('./RedirectServer');

const testStartRedirectService = options => {
  startRedirectService(options);

  expect(RedirectServer).toHaveBeenCalledWith(
    options.to,
    options.chain || false
  );
  expect(RedirectServer.prototype.listen).toHaveBeenCalledWith(
    options.port || 3000
  );
  expect(global.console.log.mock.calls).toMatchSnapshot();
};

beforeEach(() => {
  global.console = { log: jest.fn() };
});

test('should start a redirect service', () =>
  testStartRedirectService({ to: 'to-somewhere' }));

test('should start a redirect service with chain', () =>
  testStartRedirectService({ to: 'to-somewhere', chain: true }));
test('should start a redirect service with custom port', () =>
  testStartRedirectService({ to: 'to-somewhere', port: 4000 }));
