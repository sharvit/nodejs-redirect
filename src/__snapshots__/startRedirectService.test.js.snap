// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should start a redirect service 1`] = `
Array [
  Array [
    "Redirection server listening on port: 3000",
  ],
  Array [
    "Redirecting all requests to: to-somewhere",
  ],
  Array [
    "Chaining requests url is: off",
  ],
]
`;

exports[`should start a redirect service with chain 1`] = `
Array [
  Array [
    "Redirection server listening on port: 3000",
  ],
  Array [
    "Redirecting all requests to: to-somewhere",
  ],
  Array [
    "Chaining requests url is: on",
  ],
]
`;

exports[`should start a redirect service with custom port 1`] = `
Array [
  Array [
    "Redirection server listening on port: 4000",
  ],
  Array [
    "Redirecting all requests to: to-somewhere",
  ],
  Array [
    "Chaining requests url is: off",
  ],
]
`;
